File: tba-proxy.mjs
#!/usr/bin/env node
/**
 * Minimal TBA proxy for local use.
 *
 * Why:
 * - Avoid browser CORS issues
 * - Optionally keep TBA key out of the frontend (use env TBA_KEY)
 *
 * Usage:
 *   TBA_KEY=... node ./tba-proxy.mjs
 *   # or:
 *   node ./tba-proxy.mjs --port 8787
 */

import http from "node:http";
import { URL } from "node:url";

const args = process.argv.slice(2);
const portArgIndex = args.findIndex((a) => a === "--port" || a === "-p");
const port = portArgIndex >= 0 ? Number(args[portArgIndex + 1]) : 8787;

function json(res, status, obj) {
  const body = JSON.stringify(obj);
  res.writeHead(status, {
    "content-type": "application/json; charset=utf-8",
    "content-length": Buffer.byteLength(body),
    "access-control-allow-origin": "*",
    "access-control-allow-headers": "content-type,x-tba-auth-key",
  });
  res.end(body);
}

function text(res, status, body) {
  res.writeHead(status, {
    "content-type": "text/plain; charset=utf-8",
    "content-length": Buffer.byteLength(body),
    "access-control-allow-origin": "*",
    "access-control-allow-headers": "content-type,x-tba-auth-key",
  });
  res.end(body);
}

const server = http.createServer(async (req, res) => {
  try {
    if (req.method === "OPTIONS") {
      res.writeHead(204, {
        "access-control-allow-origin": "*",
        "access-control-allow-methods": "GET,OPTIONS",
        "access-control-allow-headers": "content-type,x-tba-auth-key",
      });
      res.end();
      return;
    }

    const url = new URL(req.url || "/", `http://${req.headers.host || "localhost"}`);
    if (url.pathname === "/health") {
      json(res, 200, { ok: true });
      return;
    }

    if (url.pathname !== "/api/tba") {
      text(res, 404, "Not found");
      return;
    }

    const path = url.searchParams.get("path") || "";
    if (!path.startsWith("/")) {
      json(res, 400, { error: "Missing or invalid ?path=/..." });
      return;
    }

    const key = req.headers["x-tba-auth-key"] || process.env.TBA_KEY;
    if (!key) {
      json(res, 401, { error: "Missing TBA key (provide header x-tba-auth-key or env TBA_KEY)" });
      return;
    }

    const target = `https://www.thebluealliance.com/api/v3${path}`;
    const upstream = await fetch(target, { headers: { "X-TBA-Auth-Key": String(key) } });
    const contentType = upstream.headers.get("content-type") || "application/json; charset=utf-8";
    const body = await upstream.text();

    res.writeHead(upstream.status, {
      "content-type": contentType,
      "access-control-allow-origin": "*",
      "access-control-allow-headers": "content-type,x-tba-auth-key",
      "cache-control": "no-store",
    });
    res.end(body);
  } catch (e) {
    json(res, 500, { error: String(e?.message || e) });
  }
});

server.listen(port, () => {
  // eslint-disable-next-line no-console
  console.log(`TBA proxy listening on http://localhost:${port}`);
  // eslint-disable-next-line no-console
  console.log(`Try: http://localhost:${port}/health`);
});

File: tba-proxy.py
#!/usr/bin/env python3
"""
Minimal TBA proxy for local use (no deps).

Why:
- Avoid browser CORS issues
- Optionally keep TBA key out of the frontend (use env TBA_KEY)

Usage:
  TBA_KEY=... python3 ./tba-proxy.py
  # or:
  python3 ./tba-proxy.py --port 8787
"""

from __future__ import annotations

import argparse
import json
import os
import urllib.error
import urllib.request
from http.server import BaseHTTPRequestHandler, HTTPServer
from urllib.parse import parse_qs, urlparse


def _send(handler: BaseHTTPRequestHandler, status: int, body: bytes, content_type: str) -> None:
    handler.send_response(status)
    handler.send_header("Content-Type", content_type)
    handler.send_header("Content-Length", str(len(body)))
    handler.send_header("Access-Control-Allow-Origin", "*")
    handler.send_header("Access-Control-Allow-Headers", "content-type,x-tba-auth-key")
    handler.send_header("Access-Control-Allow-Methods", "GET,OPTIONS")
    handler.send_header("Cache-Control", "no-store")
    handler.end_headers()
    handler.wfile.write(body)


class Handler(BaseHTTPRequestHandler):
    def do_OPTIONS(self) -> None:  # noqa: N802
        _send(self, 204, b"", "text/plain; charset=utf-8")

    def do_GET(self) -> None:  # noqa: N802
        try:
            u = urlparse(self.path)
            if u.path == "/health":
                _send(self, 200, json.dumps({"ok": True}).encode("utf-8"), "application/json; charset=utf-8")
                return

            if u.path != "/api/tba":
                _send(self, 404, b"Not found", "text/plain; charset=utf-8")
                return

            qs = parse_qs(u.query)
            path = (qs.get("path", [""])[0] or "").strip()
            if not path.startswith("/"):
                _send(
                    self,
                    400,
                    json.dumps({"error": "Missing or invalid ?path=/..."}).encode("utf-8"),
                    "application/json; charset=utf-8",
                )
                return

            key = self.headers.get("x-tba-auth-key") or os.environ.get("TBA_KEY")
            if not key:
                _send(
                    self,
                    401,
                    json.dumps({"error": "Missing TBA key (header x-tba-auth-key or env TBA_KEY)"}).encode("utf-8"),
                    "application/json; charset=utf-8",
                )
                return

            target = f"https://www.thebluealliance.com/api/v3{path}"
            req = urllib.request.Request(target, headers={"X-TBA-Auth-Key": str(key)})
            try:
                with urllib.request.urlopen(req, timeout=20) as resp:
                    data = resp.read()
                    content_type = resp.headers.get_content_type() or "application/json"
                    charset = resp.headers.get_content_charset() or "utf-8"
                    ct = f"{content_type}; charset={charset}"
                    _send(self, resp.status, data, ct)
            except urllib.error.HTTPError as e:
                data = e.read() if hasattr(e, "read") else b""
                ct = e.headers.get("content-type") if e.headers else "application/json; charset=utf-8"
                if not data:
                    data = json.dumps({"error": f"TBA HTTP {e.code}"}).encode("utf-8")
                    ct = "application/json; charset=utf-8"
                _send(self, e.code, data, ct)
        except Exception as e:  # noqa: BLE001
            _send(
                self,
                500,
                json.dumps({"error": str(e)}).encode("utf-8"),
                "application/json; charset=utf-8",
            )

    def log_message(self, fmt: str, *args) -> None:  # silence default logging
        return


def main() -> None:
    ap = argparse.ArgumentParser()
    ap.add_argument("--port", "-p", type=int, default=8787)
    args = ap.parse_args()

    server = HTTPServer(("127.0.0.1", args.port), Handler)
    print(f"TBA proxy listening on http://localhost:{args.port}")
    print(f"Try: http://localhost:{args.port}/health")
    server.serve_forever()


if __name__ == "__main__":
    main()

